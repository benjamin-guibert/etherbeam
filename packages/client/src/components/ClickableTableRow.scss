@use '../styles/resources/colors';

.my-clickabletablerow {
  cursor: pointer;

  &:hover {
    background-color: lighten(colors.$bg-dark-3, percentage(colors.$variant-1));
  }
}
