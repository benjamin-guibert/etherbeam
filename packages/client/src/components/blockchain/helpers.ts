import { faCoins, faFile, faWallet, IconDefinition } from '@fortawesome/free-solid-svg-icons'
import { BigNumber, ethers } from 'ethers'
import numeral from 'numeral'
import { AddressType, Token } from '../../libraries/ethereum/types'

const ETH_DECIMALS = 18
const MAX_DECIMALS = 13

export interface AmountPrintOptions {
  forceDecimals?: boolean
  decimals?: number
}

const printAmount = (amount: string, decimals: number, options?: AmountPrintOptions): string => {
  const { forceDecimals, decimals: forcedDecimals } = options ?? {}
  const decimalsCount = Math.min(forcedDecimals || decimals, MAX_DECIMALS)
  const decimalFormat = forceDecimals ? '0'.repeat(decimalsCount) : `[${'0'.repeat(decimalsCount)}]`

  return amount ? numeral(amount).format(`0,0.${decimalFormat}`) : '?'
}

export type Unit = 'wei' | 'kwei' | 'mwei' | 'gwei' | 'ether'

export const printEtherAmount = (amount: BigNumber, options?: AmountPrintOptions, unit?: Unit): string =>
  printAmount(amount ? ethers.utils.formatUnits(amount, unit) : null, ETH_DECIMALS, options)

export const printTokenAmount = (amount: BigNumber, token: Token, options?: AmountPrintOptions): string => {
  const { decimals } = token

  return printAmount(amount ? ethers.utils.formatUnits(amount, decimals) : null, decimals, options)
}

export const shortenAddress = (address: string): string => {
  const start = address.substring(0, 6)
  const end = address.substring(address.length - 4)

  return `${start}...${end}`
}

export const getAddressTypeIcon = (type: AddressType): IconDefinition => {
  switch (type) {
    case AddressType.Wallet:
      return faWallet
    case AddressType.Contract:
      return faFile
    case AddressType.Token:
      return faCoins
    default:
      return null
  }
}
