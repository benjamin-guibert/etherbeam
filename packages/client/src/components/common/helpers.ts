import { faDotCircle, faFile, IconDefinition } from '@fortawesome/free-solid-svg-icons'
import numeral from 'numeral'
import { AddressType, Token } from 'libraries/ethereum/types'
import { BigNumber, ethers } from 'ethers'

const ETH_SYMBOL = 'ETH'
const ETH_DECIMALS = 18

export interface AmountPrintOptions {
  forceDecimals?: boolean
  decimals?: number
  withSymbol?: boolean
}

const printAmount = (amount: string, symbol: string, decimals: number, options?: AmountPrintOptions): string => {
  const { forceDecimals, decimals: forcedDecimals, withSymbol } = options ?? {}
  const decimalsCount = forcedDecimals ?? decimals
  const decimalFormat = forceDecimals ? '0'.repeat(decimalsCount) : `[${'0'.repeat(decimalsCount)}]`
  const printedAmount = amount ? numeral(amount).format(`0,0.${decimalFormat}`) : '?'

  return `${withSymbol ? `${symbol} ` : ''}${printedAmount}`
}

export type Unit = 'wei' | 'kwei' | 'mwei' | 'gwei' | 'ether'

export const printEtherAmount = (amount: BigNumber, options?: AmountPrintOptions, unit?: Unit): string =>
  printAmount(amount ? ethers.utils.formatUnits(amount, unit) : null, ETH_SYMBOL, ETH_DECIMALS, options)

export const printTokenAmount = (amount: BigNumber, token: Token, options?: AmountPrintOptions): string => {
  const { symbol, decimals } = token

  return printAmount(amount ? ethers.utils.formatUnits(amount, decimals) : null, symbol, decimals, options)
}

export const shortenAddress = (address: string): string => {
  const start = address.substring(0, 6)
  const end = address.substring(address.length - 4)

  return `${start}...${end}`
}

export const getAddressTypeIcon = (type: AddressType): IconDefinition => {
  switch (type) {
    case AddressType.Contract:
      return faFile
    case AddressType.Token:
      return faDotCircle
    default:
      return null
  }
}
