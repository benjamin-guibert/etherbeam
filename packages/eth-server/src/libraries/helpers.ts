import { Contract, ContractType, Transaction, USDC_HASH, WETH_HASH } from './types'
import uniq from 'lodash/uniq'

export const sleep = (milliseconds: number, callback?: () => void): Promise<void> => {
  return new Promise((resolve) => setTimeout(callback || resolve, milliseconds))
}

export const tryAction = async <T>(
  action: () => Promise<T>,
  waitingTime: number,
  maxRetries: number,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  onError?: (error: any, count: number) => boolean | void,
  onNull?: (count: number) => boolean | void
): Promise<T> => {
  let result = undefined
  let count = 0

  while (!result) {
    try {
      result = await action()

      if (result == null) {
        count++
        if (onNull?.(count) || count > maxRetries) return null

        await sleep(waitingTime)
      }
    } catch (error) {
      count++
      if (onError?.(error, count) || count > maxRetries) return null

      await sleep(waitingTime)
    }
  }

  return result
}

export const sanitizeHash = (value: string): string => {
  if (!value) return value

  return `${value.substr(0, 2)}${value.substr(2).toLowerCase()}`
}

export const getAddressHashesFromTransaction = ({ from, to, action }: Transaction): string[] => {
  const hashes = [sanitizeHash(from), sanitizeHash(to)]
  action?.parameters.forEach((parameter) => {
    switch (parameter.type) {
      case 'address':
        hashes.push(sanitizeHash(parameter.value as string))
        break
      case 'address[]':
        ;(parameter.value as string[]).forEach((value) => hashes.push(sanitizeHash(value)))
        break
    }
  })

  return uniq(hashes)
}

export const isTransactionWatched = (transaction: Transaction, contracts: Contract[]): boolean => {
  const ignoredHashes = [WETH_HASH, USDC_HASH]
  const tokenHashes = contracts
    .filter((contract) => contract.type == ContractType.Token && !ignoredHashes.includes(contract.hash))
    .map((token) => token.sanitizedHash)
  const transactionHashes = getAddressHashesFromTransaction(transaction)

  return transactionHashes.some((hash) => tokenHashes.includes(hash))
}
