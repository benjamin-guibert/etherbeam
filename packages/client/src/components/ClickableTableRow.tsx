import React, { ReactElement, ReactNode } from 'react'
import './ClickableTableRow.scss'

interface ClickableTableRowProps {
  onClick: () => void
  children?: ReactNode
}

const ClickableTableRow = ({ onClick, children }: ClickableTableRowProps): ReactElement => {
  const onKeyUp = (event: React.KeyboardEvent<HTMLTableRowElement>): void => {
    if (event.key == 'Enter') onClick()
  }

  return (
    <tr onClick={onClick} onKeyUp={onKeyUp} className="my-clickabletablerow" tabIndex={0}>
      {children}
    </tr>
  )
}

export default ClickableTableRow
