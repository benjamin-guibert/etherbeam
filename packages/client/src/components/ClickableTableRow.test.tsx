import React from 'react'
import { mount, ReactWrapper } from 'enzyme'
import { act } from 'react-dom/test-utils'
import { clearMocks } from '../../tests/helpers'
import ClickableTableRow from './ClickableTableRow'

describe('<ClickableTableRow />', () => {
  const onClick = jest.fn()

  let component: ReactWrapper

  describe('Default', () => {
    beforeAll(() => {
      clearMocks()
      component = mount(
        <table>
          <tbody>
            <ClickableTableRow onClick={onClick}>
              <td></td>
            </ClickableTableRow>
          </tbody>
        </table>
      )
    })

    it('should render', () => {
      const tr = component.find('tr')
      expect(tr.hasClass('my-clickabletablerow')).toBeTruthy()
      expect(tr.find('td').exists()).toBeTruthy()
    })

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('On click', () => {
    beforeAll(() => {
      clearMocks()
      component = mount(
        <table>
          <tbody>
            <ClickableTableRow onClick={onClick}>
              <td></td>
            </ClickableTableRow>
          </tbody>
        </table>
      )

      act(() => {
        component.find('tr').simulate('click')
        component.update()
      })
    })

    it('should call onClick', () => expect(onClick).toBeCalledTimes(1))
  })

  describe('On key up (enter)', () => {
    beforeAll(() => {
      clearMocks()
      component = mount(
        <table>
          <tbody>
            <ClickableTableRow onClick={onClick}>
              <td></td>
            </ClickableTableRow>
          </tbody>
        </table>
      )

      act(() => {
        component.find('tr').simulate('keyUp', { key: 'Enter' })
        component.update()
      })
    })

    it('should call onClick', () => expect(onClick).toBeCalledTimes(1))
  })

  describe('On key up (other)', () => {
    beforeAll(() => {
      clearMocks()
      component = mount(
        <table>
          <tbody>
            <ClickableTableRow onClick={onClick}>
              <td></td>
            </ClickableTableRow>
          </tbody>
        </table>
      )

      act(() => {
        component.find('tr').simulate('keyUp', { key: 'a' })
        component.update()
      })
    })

    it('should not call onClick', () => expect(onClick).toBeCalledTimes(0))
  })
})
