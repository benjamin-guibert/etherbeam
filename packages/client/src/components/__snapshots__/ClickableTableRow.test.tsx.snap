// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClickableTableRow /> Default should match snapshot 1`] = `
<table>
  <tbody>
    <tr
      class="my-clickabletablerow"
      tabindex="0"
    >
      <td />
    </tr>
  </tbody>
</table>
`;
